### ISSUE TITLE
TypeError: Cannot read property 'lastChange' of undefined

### ISSUE THREAD

**Issue by bisol** [paulmillr/chokidar#546]
Got this error twice, not sure how to reproduce. This does not seems stop the watcher, and the same code runs fine on other systems. Node v5.12.0 on Ubuntu 10.04 with Chokidar 1.6.1

The app is configured to monitor a root dir with many subdirs (max depth = 1), each with many files (about 5Tb of 50kb to 30Mb files).

The error occurs on index.js at line 309:
      if (now - this._pendingWrites[path].lastChange >= threshold) {

Chokidar is configured thus:

watcher = chokidar.watch(path, {
  ignoreInitial: true,
  followSymlinks: false,
  awaitWriteFinish: true,
  ignorePermissionErrors: true,
  depth: 0
})
watcher.on('addDir', function (path) {
  watcher.add(path)
})

**Comment by es128** [comment]
Looks like a bug occurring when the same file gets an event multiple times within the awaitWriteFinish polling interval.

If you can reproduce it reliably, try applying this patch and repeating the actions that would cause the error. Let me know if it works to prevent it.

```diff
diff --git a/index.js b/index.js
index e23a643..4b3fa0d 100644
--- a/index.js
+++ b/index.js
@@ -295,8 +295,8 @@ FSWatcher.prototype._awaitWriteFinish = function(path, threshold, event, awfEmit

   var awaitWriteFinish = (function (prevStat) {
     fs.stat(fullPath, function(err, curStat) {
-      if (err) {
-        if (err.code !== 'ENOENT') awfEmit(err);
+      if (err || !(path in this._pendingWrites)) {
+        if (err && err.code !== 'ENOENT') awfEmit(err);
         return;
       }
```

**Comment by bisol** [comment]
Thanks for the quick response.
This happened two times yesterday, and again just now (after running for about 4 hours).
This app runs out of memory and restarts itself (using recluster) every few hours, so that may be the cause

**Comment by es128** [comment]
It still indicates a flaw in the code that my proposed patch should resolve, but it would be nice to get some validation on that if you have the ability to try it.

**Comment by bisol** [comment]
Yes, I should have mentioned I update my app with your patch and will redeploy it tomorrow (it's getting late here, and I don't own that server)

**Comment by es128** [comment]
Resolved by #644 and soon to be published in v2.0.1
